usr/lib/common/asn1.c

```diff
@@ -3473,3 +3574,806 @@ CK_RV ber_decode_DHPublicKey(CK_BYTE *data,
+            memcpy(buf + offset, buf2, len);
+            offset += len;
+            free(buf2);
+            buf2 = NULL;
+        }
+
+        /* seed */
+        rc = ber_encode_BIT_STRING(FALSE, &buf2, &len,
+                                (CK_BYTE *) seed + sizeof(CK_ATTRIBUTE),
+                                seed->ulValueLen, 0);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_encode_BIT_STRING of seed failed\n");
+            goto error;
+        }
+        if (buf2 != NULL) {
+            memcpy(buf + offset, buf2, len);
+            offset += len;
+            free(buf2);
+            buf2 = NULL;
+        }
+
+        /* tr */
+        rc = ber_encode_BIT_STRING(FALSE, &buf2, &len,
+                                (CK_BYTE *) tr + sizeof(CK_ATTRIBUTE),
+                                tr->ulValueLen, 0);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_encode_BIT_STRING of (tr) failed\n");
+            goto error;
+        }
+        if (buf2 != NULL) {
+            memcpy(buf + offset, buf2, len);
+            offset += len;
+            free(buf2);
+            buf2 = NULL;
+        }
+
+        /* s1 */
+        rc = ber_encode_BIT_STRING(FALSE, &buf2, &len,
+                                (CK_BYTE *) s1 + sizeof(CK_ATTRIBUTE),
+                                s1->ulValueLen, 0);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_encode_BIT_STRING of (s1) failed\n");
+            goto error;
+        }
+        if (buf2 != NULL) {
+            memcpy(buf + offset, buf2, len);
+            offset += len;
+            free(buf2);
+            buf2 = NULL;
+        }
+
+        /* s2 */
+        rc = ber_encode_BIT_STRING(FALSE, &buf2, &len,
+                                (CK_BYTE *) s2 + sizeof(CK_ATTRIBUTE),
+                                s2->ulValueLen, 0);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_encode_BIT_STRING of (s2) failed\n");
+            goto error;
+        }
+        if (buf2 != NULL) {
+            memcpy(buf + offset, buf2, len);
+            offset += len;
+            free(buf2);
+            buf2 = NULL;
+        }
+
+        /* t0 */
+        rc = ber_encode_BIT_STRING(FALSE, &buf2, &len,
+                                (CK_BYTE *) t0 + sizeof(CK_ATTRIBUTE),
+                                t0->ulValueLen, 0);
+        if (rc != CKR_OK) {
+            TRACE_DEVEL("ber_encode_BIT_STRING of (t0) failed\n");
+            goto error;
+        }
+        if (buf2 != NULL) {
+            memcpy(buf + offset, buf2, len);
+            offset += len;
+            free(buf2);
+            buf2 = NULL;
+        }
+
+        /* (t1) Optional bit-string of public key */
+        if (t1 && t1->pValue) {
+            ber = ber_alloc_t(LBER_USE_DER);
+            rc = ber_put_bitstring(ber, t1->pValue, t1->ulValueLen * 8, 0x03);
+            rc = ber_flatten(ber, &val);
+
+            ber_encode_CHOICE(FALSE, 0, &buf2, &len, (CK_BYTE *)val->bv_val,
```

Comment:
error checking ?

Reply:
Good point ... and there are a couple of other code pieces with the same logic.